FormBuilder Internationalization (i18n)

samuelsov opened the pull request:
Overview
----------------------------------------
Internationalization support for FormBuilder.

Before
----------------------------------------
Translation is working if the translation is provided in core using .mo files but not for custom text.

After
----------------------------------------
Translation is possible using the `civicrm_translation` table and a new `civicrm_translation_source` table to store what could be translated.

To test:
- enable Multilingual mode + disable Asset Caching + enable extension kamlanguage, i.e. Language switcher (not required but easier)
- create or edit a form builder and add custom strings as label, placeholder....
- upon saving the form, the table `civicrm_translation_source` is populated with the new strings
- to translate, go to the menu Administer -> Localisation -> Translations to XXX
<img width="821" height="479" alt="Capture d’écran du 2025-07-22 16-15-30" src="https://github.com/user-attachments/assets/0798060c-c009-4bbb-b30e-d9106e735b45" />
- then visit the form builder page and switch the language (using the kamlanguage menu)
- Result: the form should be translated

Technical Details
----------------------------------------
Instead of adding `ts` to `.aff.html` file, which causes issues in the GUI, we now extract any strings from the `.aff.html` files that should be localizable (title, button, text...) and put them in a new table `civicrm_translation_source`.

When saving an afform in the GUI, the extractor is called to save the new strings as data in `civicrm_translation_source`.

Using a specialized SearchKit based on a SQL view, we are able to do inline edit translations of string data in the `civicrm_translation` table. It works even if there are no translations yet in the `civicrm_translation` table.

The `ts` function will now look in the `civicrm_translation` first and then fallback to the regular transifex mo files. This allows one to provide a default for packaged afform and yet still be able to change them.

Comments
----------------------------------------
Discussed but excluded from scope (and will lead to future PRs):
- the title of the afform is not yet translated
- the asset cache needs to be manually cleaned-up after adding some translation (or disabled)
- while the database supports context (i.e. which afform/field the translation applies to), it is not yet taken into account which means that the same source string will be translated the same across all afform
- some database changes due to the use of `civicrm_translation` impact current Message translation intended to support translations of strings from other entities
- the existing word replacement feature is very similar and could/should probably use this new table
- remove all `ts` from existing afform - some needs extra change in the tags or angular

samuelsov committed:
Formbuilder translation: new TranslationSource table

colemanw committed:
Add upgrader for TranslationSource table

totten commented on ext/afform/core/Civi/Afform/Translator.php:
```diff
@@ -0,0 +1,90 @@
+- [af-title] */
+
+    $changeSet = \Civi\Angular\ChangeSet::create('translate')
+      ->alterHtml(';.aff.html;', function (\phpQueryObject $doc) {
```
Tighter regex?

```suggestion
      ->alterHtml(';\.aff\.html$;', function (\phpQueryObject $doc) {
```

samuelsov replied:
Thanks @totten this is fixed as suggested.

totten commented on ext/afform/core/Civi/Afform/Translator.php:
```diff
@@ -0,0 +1,90 @@
+
+  // TODO: make this a generic class
+  public function translateContent(&$item) {
+    $item->textContent = ts($item->textContent);
```
IIRC, `ts()` and `_ts()` are basically the same. Except that linters may complain about dynamic lookupers (`ts($foo)`). So `_ts($foo)` is suggested for dynamic lookups.

samuelsov committed:
Upgrader for initializating afform i18n sources

samuelsov committed:
Cleanup and suggestion from Tim

shaneonabike committed:
Selectors for auto-ts in one place

colemanw committed:
Tests - Decouple LocaleTestTrait from Api3

Allows Api4 tests to use this trait.

colemanw committed:
Add separator in localization menu

Add separator between the existing items and the new
translation search displays.

colemanw committed:
Translation - Generate Search/Display/Afform per language

Creates an editable search display for each language,
enabling source strings to be translated.

colemanw commented:
@samuelsov @shaneonabike after some expermintation I found that a specialized api4 entity/view was not necessary. I fixed a couple bugs in SK editable and was able to get the job done by autogeneratng 1 search/displayafform per language. You'll find them in the Localization menu :)

shaneonabike commented:
Yeah I noticed those as well @colemanw. I fixed a bunch of that on the plane home but I just have to merge it with your new things! I also caught te missing placeholders too. Thanks for your help!

colemanw commented:
@samuelsov @shaneonabike the test fail is:
```
In AfformSaveTrait.php line 107:
                                                                               
  [ArgumentCountError]                                                         
  Too few arguments to function CRM_Afform_Upgrader::saveTranslations(), 1 pa  
  ssed in /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/ex  
  t/afform/core/CRM/Afform/Upgrader.php on line 115 and exactly 2 expected     
                                                                               

Exception trace:
  at /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/ext/afform/core/Civi/Api4/Utils/AfformSaveTrait.php:107
 CRM_Afform_Upgrader->saveTranslations() at /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/ext/afform/core/CRM/Afform/Upgrader.php:115
 CRM_Afform_Upgrader->upgrade_1004()
 ```

totten committed:
AfformTranslator - When swapping translated text, preserve outer HTML

totten committed:
(REF) AfformSaveTrait - Extract the StringScanner class

MOTIVATION: Make it easier to write a unit-test that calls the built-in string-scanner.

BEFORE: The `AfformSaveTrait` has several helper functions to perform scanning. The
results are stored in a buffer ($this->stringTranslations).

AFTER: The `AfformSaveTrait` uses the `StringScanner`. This class has all the same
helper functions and a similar buffer.

totten committed:
(NFC) StringScanner - Add FIXME

totten committed:
TranslatorTest - Add headless test for scanning=>saving=>using translations in afform

totten committed:
(NFC) TranslationSource.php, I18n.php, JS.php, Translate.php

totten committed:
(NFC) OptionValueLinksProvider.php, AfformTranslateTrait.php

samuelsov commented:
@totten ok, great! Shane and I fixed a few issues yesterday. I'm working on merging them with the change you made.

samuelsov committed:
Afform translation - Recursive defn selector

Co-Authored-By: Shane <shane@symbiotic.coop>

samuelsov committed:
Afform translator - Missing trim for getting html translation

samuelsov commented:
I need to do a few more tests but it's seems pretty good to me. The only missing translation on a afform is the title of the form. It's populated in the `civicrm_translation_source` but we miss the part to translate it when loading the afform.

totten committed:
StringScanner - Fix PHP warning

shaneonabike commented:
@totten I was thinking the same. Originally, I had setup a callback when the result was discovered, but it got complicated so we abandoned it and didn't mention that it needed work.

If you suggest where I could create a new class to do this exploring I'm all ears? We would need this to be available for Searchkit in the future also b/c that would be the next iteration?

totten committed:
TranslatorTest - Add more varied examples of <br>

totten committed:
StringScanner - Align more with Translator (re: serialization of inner HTML)

totten committed:
(REF) Convert TranslatorTest to MockTranslatorTest with many small files

totten committed:
StringScanner, Translator - Fix error

Presents messages like:

  foreach() argument must be of type array|object, null given

via

/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/ext/afform/core/Civi/Afform/StringScanner.php:77
/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/ext/afform/core/Civi/Afform/StringScanner.php:53
/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/rubobaquero/phpquery/phpQuery/phpQuery.php:1089
/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/rubobaquero/phpquery/phpQuery/phpQuery/phpQueryObject.php:2928
/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/ext/afform/core/Civi/Afform/StringScanner.php:55
/home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/ext/afform/mock/tests/phpunit/Civi/AfformMock/MockTranslateTest.php:101

totten committed:
Afform - Consolidate Translator and StringScanner

NOTE: There were a couple discrepancies between `Translator` and `StringScanner`, which
are resolved as follows:

- HTML Entities: `Translator` was performing a round of decode/encode for HTML entities,
  which StringScanner did not. `Translator` cared because it was using a feed of
  translations from SQL... which were _stored_ with extraneous HTML entities.

  I suspect that getTranslationSources() was actually broken for handling for
  any strings containing markup (*not just afform strings*). This update
  should be a more general fix.

- Angular Directive Bypass: `StringScanner` would bypass anything that looked like
  it included an Angular directive (`{{`). I'm not certain if it's good or bad,
  but it seems reasonable enough. If we're doing it, we should consistently.

  The update should apply the same bypass during both scan+translate.

totten commented:
> (@shaneonabike) If you suggest where I could create a new class to do this exploring I'm all ears? We would need this to be available for Searchkit in the future also b/c that would be the next iteration?

I'm not sure I see the difficulty, so I took a whirl and produced https://github.com/samuelsov/civicrm-core/commit/29be919fcd8ab93bb1330c7b67f6a1886a1df8c9

(Before trying that, I wanted some more test-coverage. So full proposal is https://github.com/samuelsov/civicrm-core/compare/afformi18n...totten:civicrm-core:afformi18n-consolidate?expand=1. The other big thing in that branch is splitting the TranslatorTest into separate files. This makes it easier to copy-in more complex examples.)

> (@samuelsov) html content is not always translated because some mismatch (accentued character are escaped in Translator but not in StringScanner)

Ah, yeah. I noticed something similar while taking a stab at consolidating the classes. My feeling is that `Translator` is compensating for a bug that's somewhere else. If that bug is fixed, then `Translator` and `StringScanner` become more similar (as in the proposed `StringVisitor`.) https://github.com/samuelsov/civicrm-core/commit/29be919fcd8ab93bb1330c7b67f6a1886a1df8c9

To see what I mean...

* Forget Afform for a moment. Just think about the  APIs (e.g. `Translation::create()` and `TranslationSource::create()`) and SQL tables (`civicrm_translation`, `civicrm_translation_source`).
* Any time you write to these tables via API, it goes through `HTMLInputCoder`:
     * If you say `setValues(['string' => 'Hello<br/>'])`...
     * Then SQL stores... `string='Hello&lt;br/&gt;'`.
* We normally don't notice because API's have automatic-decoding.
* But for `ts()` integration, it uses an optimized read SQL (`CRM_Core_BAO_TranslationSource::getTranslationSources()`) ==> no automatic-decoding. 
* If we teach `getTranslationSources()` to apply the decoding, then it should help handling of all `ts()` calls involving data from `Translation`/`TranslationSource` (*including, but not limited to, afform*).

totten committed:
afform_mock - Fix for current translations

I had an old copy of the `l10n` folder. This is more current - and matches what CI has.

samuelsov committed:
Afform transaltion - Translating title + fix for empty translation

samuelsov commented:
@totten @colemanw Another problem that we have is that when we add data in the translation source table, afform is not aware of it until we clear the cache or we disable asset caching. Should we force refresh the translation caching (asset_builder?) after an insert in the table ? otherwise, what do you think would be the best way to deal with that ?

totten commented:
Oh, yeah, that's a tricky one @samuelsov. I'm not really sure the best way. Some brainstorming....

* I guess it's hard to say without some quantitative assumptions -- like how many forms exist; how many strings; how many locales; how often the strings are edited; how many concurrent users access the forms. With lower numbers, nobody worries about doing full flush. With higher numbers, you start thinking about incremental flush. 

    (What's reasonable? Maybe we imagine 100 forms and 4 languages? 100x4?)

* Qualitatively, the design of the translation UX makes a difference as well:

    * Editing strings 1-by-1.... In one "session" where a translator is active, they edit something every 30 seconds. Worst-case, that's flushing 100x4 cache-items every 30 seconds. In that period, caching would be seem to be  almost counterproductive?
    * Editing strings in batch (eg CSV/POT import)... You don't submit batches as often. So it's OK to be more heavy-handed with the flushes.

* It would be possible to keep an index that tracks string-usage in detail -- and whenever there's a change, use the index to issue targeted flushes. My guess is that it probably _scales_ better... and feels automatic... but also _adds fixed overhead_ (re: code-complexity and/or build-time performance). Dunno if it's worth it.

* Maybe a middle-ground is to flush by locale?
    * Example A: Use folders with explicit flushing
        1. Update `AssetBuilder` to organize cache-files into a locale folders (`angular-modules.XXX.json` ==> `fr_CA/angular-modules.XXX.j.son`).
        2. Whenever you edit a string in `fr_CA` , then it should flush the entire `fr_CA/` folder.
        3. (*Hrm, but you can't flush client-side caches?*)
    * Example B: Use last-modification time with implicit flushing
        1. Update `civicrm_translation` to track a `modified_date`.
        2. This allows you to determine when a locale was last edited, eg
            ```sql
            SELECT MAX(modified_date) FROM civicrm_translation WHERE locale = 'xx_XX' AND status_id=1;
            ```
        4. When [computing the digest of asset-inputs](https://github.com/civicrm/civicrm-core/blob/7ef0e179d2dcb5e370f722cae5c1ac80bd4b16d9/Civi/Core/AssetBuilder.php#L288-L297), include the last-modification time of the active locale.

* Or maybe we just make it explicit. If you are a translator, then you have a couple extra options available:
    * `Translator Mode` = `On || Off`
        * _Your user will stop using caches. (But everybody else continues using caches.)_
    * `Translation Timestamp: Last edit at 4:15PM. Last flush at 4:00PM. [Flush now]`
        * _You're notified if the caches are out-of-date, and you have an option to flush._

samuelsov commented:
@totten For batch translation, it could makes sense to have an explicit cache flush when the translator is done with the current work. There is no need for real time in that case if the UI is clear like you mentioned (`Last edit at 4:15PM. Last flush at 4:00PM. [Flush now]`)

However, if we aim for in-place single translation (e.g. in a popup), it seems a bit too much to ask for a flush after a small translation.

I think I like the compromise of a by language implicit flushing based on last-modification (your example B). Like for Smart groups, we could have minimum duration of cache. For small site, we could set this to 0 so it refresh any time a change is done and for bigger site, we could define a minimum number of minute before the cache expires.
This is also compatible with an explicit `Flush now` if we want to let translator test their batch translation without having to wait for too long.

Although if I am to do this, I'm not quite sure how the digest works. Can't we just add the ` last edited` check in `CRM_Core_I18n::getTranslationReplacements` ?

samuelsov committed:
Afform transaltion - fix for CiviCampaign dashboard

samuelsov committed:
Afform translation - more performant database structure

Co-authored-by: Shane <shane@symbiotic.coop>

samuelsov committed:
afformi18n pushed to CiviCRM 6.6

samuelsov committed:
afformi18n init translations sources when enabling i18n

samuelsov committed:
afformi18n - translation entity_table must be default to null for the SK to work

samuelsov commented:
Ok, I think we are ready for a review here. We have a functional afform multilingual translation which should work as a base for other entity (e.g. SK) translation in the near future.

totten committed:
TranslationSource.entityType.php - Fix reference to non-existent column

totten committed:
Upgrader - createEntityTable should generate sensible error if schema is missing

totten committed:
MockTranslateTest - Fix schema-reset to match current schema

totten committed:
(fixup) afformi18n pushed to CiviCRM 6.6

totten commented:
Updating the `TranslationSource` metadata resolved a bunch of failures (400=>60). A few more tweaks reduced errors to 46.

It's been a while, so my memory is a bit cold, but... there seems to be a mismatch where:

* The schema for `TranslationSource` [requires](https://github.com/civicrm/civicrm-core/blob/e24cdd61c302b22614380fdad67e34bad12a0f9f/schema/Core/TranslationSource.entityType.php#L37) `entity`, `context_key`, `source_key`.
* The MockTranslateTest [populates](https://github.com/civicrm/civicrm-core/blob/e24cdd61c302b22614380fdad67e34bad12a0f9f/ext/afform/mock/tests/phpunit/Civi/AfformMock/MockTranslateTest.php#L143) with just `source`.

And I vaguely recall that this schema was evolving, but I'm not sure where it was supposed to land. However, it is consistent with the [test-failures in MockTranslateTest](https://test.civicrm.org/job/CiviCRM-Test/93328/testReport/%28root%29/Civi_AfformMock_MockTranslateTest/testTranslate_with_data_set__tStaticText_/).

Perhaps it needs an update for populating `TranslationSource`?

samuelsov commented:
@totten Indeed, entity should be `afform` in this case. I feel that `context_key` and `source_key` should be populated automatically (in the BAO or api call?) but right now, it's seems to be required / done manually. We use a function to compute them `\CRM_Core_BAO_TranslationSource::createGuid` but it's not implicit yet.

mlutfy commented:
Mysterious test fail:

```
ok 2472 - api_v3_SystemTest::testSystemGet
not ok 2473 - Failure: api_v3_SystemTest::testSystemUTFMB8Conversion
  ---
  message: 'Failure in api call for System utf8conversion:  DB Error: unknown error'
  severity: fail
  ...
```

samuelsov committed:
afformi18n - fixing unittest

dontub commented:
How can extensions with afforms provide translations for those?

shaneonabike commented:
> I have tested in the on line demo. Impressive ! That is a major step forward the multi language. I saw that the form title was also translated !

@Guydn if you have time to stress test and create different forms with various things that could be help. We have done quite a bit but any additional testing would be extremely helpful. We are pretty excited about this development

> How can extensions with afforms provide translations for those?

The way that this works is that it hooks into Afform right now so during upgrades it scans _all_ FormBuilder forms and integrates them into the language translation system _if multilingual is installed_. After that, if you add a new FormBuilder and save it then it's added as well.

Any strings that are within an extension exported to a _.po_ are used **first** as a means of translation for any string that is within a FormBuilder. So basically, if you ship an extension with a po then it will just kind of work.

Guydn commented:
I also see that exposed filters of a search form are translated as well. Obviously columns headers are not.

Why should we disable asset caching (what does it mean ?) ?

It is not clear for me in which language should the form be created. I suppose there should be only 1 language for form creations. I started to create the form in english then i created the translations to french. When I switch to french all fields are translated. Fine.
If I create the form in French, I set the custom fields (ex: marital status, marriage date, most important issue), i renamed them in french, I use the translations to english. If I switch to english, fields labels are not changed.

shaneonabike commented:
> It is not clear for me in which language should the form be created

Yes we need more docs around that. Essentially, no matter what language you have created the form in is the language the strings are translated in. We did this _because_ there are plenty of folks whom their primary language is not English.

So we have:

+ Primary language: Spanish
+ Secondary language: English

Using the site, Spanish user creates a Form and saves so the strings on the fields are saved as Spanish. Using the translation interface they would then translate `into the secondary, tertiary, ... language`. But it's heavily based on the present language of your interface I believe.

> also see that exposed filters of a search form are translated

`Any field` in FormBuilder will be translated including options (if they are customized. Table headers are not translated because that's embedding a Searchkit. That's the next stage if we can get this into core ASAP.

> If I create the form in French, I set the custom fields (ex: marital status, marriage date, most important issue), i renamed them in french

Does this mean that you created a form in French or in English? You need to change your interface language to French, and build the form that way. Then you will get strings ready to be translated to English.

If that doesn't resolve your issue can you post your actual FormBuilder in a code snippet for testing, and provide further details.

samuelsov commented:
@Guydn 

> I also see that exposed filters of a search form are translated as well. Obviously columns headers are not.

This is one of the next steps, once this PR is committed.

> Why should we disable asset caching (what does it mean ?) ?

When you go to Administer -> System Settings -> Debugging and error handling, you need to set the field Asset Caching to disable (in civicrm/admin/setting/debug?reset=1). This is a bug we have with the caching of the translation. It will be solved eventually but we don't want to bloat this PR too much.

> It is not clear for me in which language should the form be created. I suppose there should be only 1 language for form creations. I started to create the form in english then i created the translations to french. When I switch to french all fields are translated. Fine. If I create the form in French, I set the custom fields (ex: marital status, marriage date, most important issue), i renamed them in french, I use the translations to english. If I switch to english, fields labels are not changed.

This should work. It shouldn't matter the language you are when you do the translation but you need to be in the good Search Kit. If you translate from French to English, you should be in the page Translation to English.

My guess is that asset caching is still enabled in your case so you need to disable it.

dontub commented:
Thanks for the explanations @shaneonabike, @mlutfy. I was wondering how civistrings would extract the translatable strings if no `ts()` is used. Meanwhile  I've seen that you're working on it https://github.com/civicrm/civistrings/pull/18

Guydn commented:
I try to explain what I did.
Primary language English, secondary language French
Asset caching is disabled.
language is english. 
Step 1: I create a form (individual), I add the email block and the 3 custom fields "Marriage date", "Marital status", "Most important issue". The are renamed in the form with these labels.
Switch to French: in Formbuilder, the labels First name, last name and email are now in French. The 3 other labels are the same.
Switch to english
Step 2: I go to the screen "translations to French" and I add the transation "date de mariage 1", "statut marital 1", "sujet important 1".
Step 3: I run the form, all labels are in english
Switch to French
Step 4: I run again the form, all labels are in french: first name, last name, email and the 3 custom fields that were given a translation
Switch again to english, the form is displayed in english.
Switch back to french
Step 5: go to formbuilder (edit the form). First name, last name and email are in french but the 3 custom fields are in english. You would expect that the translations are retrieved . However if I run the form, all labels (custom fields included) are in french. 
Step 6: create a form (language is still switched to french) Rename directly in the form the 3 custom fields to "date de mariage 2", "statut marital 2", "sujet important 2".
Step 7: Go to the screen "translations to english". I find the labels "date de mariage 2", "statut marital 2", "sujet important 2" above that I can translate to english. I give as translations: marriage date 2, marital status 2, most important issue 2.
Step 8: run the form. all fields are in french.
Switch to english
Step 9: If i run the form. OK for standard fields; But the custom fields are kept in French, i.e. date de mariage 2, statut marital 2, sujet important 2. I would have expected the translations marriage date 2, marital status 2, most important issue 2
There is something that I am missing.

samuelsov commented:
@Guydn Thanks for the detailed testing log.

First, I want to confirm that you saw the expected behavior.

The editor shows you the source string as in the `civicrm_translation_source` table. If you keep the default value, the source does not come from the `civicrm_translation_source` table so the translation is working as usual using the .mo files.

I understand that it can be confusing but it was too big of a step to allow translation directly from the afform.

We'll try to come up with a better implementation in a subsequent PR. We might want to show the source or the translation and indicate which one it is (if we change the source, a new translation needs to be provided so it's important to know the difference). Let's continue the discussion in mattermost so we can find a better way to do this.

Guydn commented:
ok @samuelsov that is already a huge step ! I fully understand the reasons :-).

totten committed:
Re-target Afform translation (6.6.alpha => 6.7.alpha)

totten committed:
Upgrader - Add civicrm_translation.source_key and its index

totten commented:
Doing basic `r-run` on a clean installation seemed good. 🎉 The fact that it's been running on some client sites is huge.

When I did an `r-run` with an upgraded site, I noticed some issues (e.g. the new column `civicrm_translation.source_key` needs to be created). So I added a few tweaks for the upgrader.

At any rate, I'm +1 on merging for the current 6.7.alpha (after the tests pass again). Once we get this big piece in place, I'm sure more incremental revisions can follow. :) 

This is a big change. Congrats to @samuelsov and @shaneonabike and everyone else for pushing it through!

seamuslee001 commented:
Test fail seems to be an issue

```
Civi\AfformMock\MockTranslateTest::testTranslate with data set "tTranslatedDefn" ('tTranslatedDefn')
Rendered HTML should have translations applied
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
   <fieldset af-fieldset="Individual1" class="af-container" af-title="Individual 1">\n
     <div class="af-container">\n
       <div class="af-container af-layout-inline">\n
-        <af-field name="first_name" defn="{input_attrs: {placeholder: '(Your given name)', size: '30', label: 'Pr\u00e9nom'}, label: 'Given Name', required: false, name: 'first_name', data_type: 'String', fk_entity: null, dfk_entities: null, options: false, input_type: 'Text', help_pre: null, help_post: null, search_range: false, template: '~/af/fields/Text.html'}"></af-field>\n
+        <af-field name="first_name" defn="{input_attrs: {placeholder: '(Your given name)', size: '30', label: 'First Name'}, label: 'Given Name', required: false, name: 'first_name', data_type: 'String', serialize: null, fk_entity: null, dfk_entities: null, options: false, input_type: 'Text', help_pre: null, help_post: null, search_range: false, template: '~/af/fields/Text.html'}"></af-field>\n
       </div>\n
     </div>\n
   </fieldset>\n
   <button class="af-button btn btn-primary" crm-icon="fa-check" ng-click="afform.submit()">Soumettre</button>\n
 </af-form>'

/home/homer/buildkit/build/build-1/web/core/ext/afform/mock/tests/phpunit/Civi/AfformMock/MockTranslateTest.php:101
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2538
```

totten commented:
```
Civi\AfformMock\MockTranslateTest::testTranslate with data set "tTranslatedDefn" ('tTranslatedDefn')
```

Feels like an interaction with some other recent merge. The test passes on the PR-branch but fails with merged master+PR.

totten commented:
@colemanw It looks like an interaction with https://github.com/civicrm/civicrm-core/pull/33447. Any ideas?

<details>
<summary>Debugging</summary>

Get a list of recent merges

```
3c4c337338 Merge pull request #33468 from colemanw/boilerplateGone
3dff8d0d4b Merge pull request #33453 from colemanw/standaloneSettingsCleanup
cb7fa2d355 Merge pull request #33424 from colemanw/afformFilterClause
6be6351868 Merge pull request #33464 from vingle/searchkit-table-button-align
e4ba3effa7 Merge pull request #33454 from colemanw/searchGroupByEditable
e32b6c83d3 Merge pull request #33458 from colemanw/afformMultipleHiddenValues
2a0bd822f5 Merge pull request #33465 from totten/master-stripe-cc-block
8f9bcb64bd Merge pull request #33450 from totten/master-payproc-edit
9ff578a6d1 Merge pull request #33447 from colemanw/getFieldsCallback
66e327ba38 Merge pull request #31326 from jmcclelland/output-iframe-code
db8093c543 Merge pull request #33457 from webmaster-cses-org-uk/patch-2
d8fc28e375 Merge pull request #33461 from colemanw/afformLinkPerm
a4101a5328 Merge pull request #33374 from ufundo/searchkitreports-permission
697c109b89 Merge pull request #33456 from vingle/datepicker-darkmode2
e8ef8a6dc3 Merge pull request #33455 from vingle/datepicker-darkmode
78cd490940 Merge pull request #33324 from vingle/AF-details-inline-layout
2a0343258b Merge pull request #33451 from colemanw/match
e8927f9bcf Merge pull request #33444 from vingle/wp-radio-border
```

Do a manual bisection - checkout the recent-merge then add-on this PR and see if it the test passes or fails

```
phpunit9  tests/phpunit/Civi/AfformMock/MockTranslateTest.php  --stop-on-failure
````

My bisection results:

```
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 3c4c337338
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 3dff8d0d4b
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge cb7fa2d355
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 6be6351868
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge e4ba3effa7
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge e32b6c83d3
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 2a0bd822f5
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 8f9bcb64bd
[f] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 9ff578a6d1
[p] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 66e327ba38
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge db8093c543
[p] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge d8fc28e375
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge a4101a5328
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 697c109b89
[p] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge e8ef8a6dc3
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 78cd490940
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge 2a0343258b
[ ] git checkout master ; git branch -D merge-tmp ; git checkout afformi18n -b merge-tmp ; git merge e8927f9bcf
```

It seems that the failure starts with 9ff578a6d1.

</details>

shaneonabike commented:
chances that it is because when we use this new integration you have to disable caching or it doesn't work. This is another thing we have to deal with down the line obviously.

totten committed:
Update example: tTranslatedDefn.rendered.html

totten commented:
OK, the newer failure appears in the same test-class... but is actually a different thing. (It's an interaction with 4f1c71f6a3.) For this, the test needs an update. But it's gonna be a lot easier to do that with a rebase....

colemanw commented:
Congratulations @samuelsov @shaneonabike! Great accomplishment getting this finished 🎉🍾

shaneonabike commented:
Thanks but you gotta open some champagne for yourself too! I appreciate that help at the sprint from both @colemanw @totten and others who helped us brainstorm, and add certain elements. @samuelsov and I are already generating a larger list of work that will need to be tackled to fully integrate this :firecracker:
